plugin: add multifactor priority plugin

cmoussa1 opened the pull request:
This is a very early first-cut, work-in-progress PR that adds multi-factor priority calculation to flux-accounting. It is still missing a couple of integral pieces, but I thought it might be best to post it early for the following reasons:

- Since this is a somewhat important milestone for flux-accounting as a whole, I wanted to make sure I have been headed in the right direction so far with my plugin
- I think I might benefit from some collaboration and feedback on the pieces that are still missing from this plugin (testing strategy, the data structure in the plugin, managing users who belong to more than one bank, etc.)

---

#### Background

A multifactor priority plugin is responsible for calculating a job's priority using a number of different factors that may help or hurt a job's overall priority. This PR starts with adding just two factors: **fairshare** and **urgency**. The intent is that once this PR becomes mergeable, we can start to add additional factors in follow-up PR's to create a more complete and solidified priority calculator.

The **fairshare** factor represents the ratio between the amount of allotted resources and the amount of resources actually consumed. It should range between 0.0 and 1.0, where 0.0 represents the lowest possible value (indicating a user who is using way more resources than allocated) and 1.0 represents the highest. When a user first gets added to the database, their fairshare value should be 0.5, which should mean the user is using exactly the amount of resources they were allocated.

The **urgency** factor is similar to the **nice** factor in Slurm in that it allows users to prioritize their own jobs with an integer in the range [1,16], where 16 is the highest priority and 1 is the lowest.

Together, these two factors make up a priority calculation using the following formula:

```
priority = | (fshare_weight * fshare_factor) + (urgency - 16) |
```

where:

- **fshare_weight**: represents the integer weight that fairshare has on the overall priority calculation
- **fshare_factor**: represents a user's fairshare value stored in the flux-accounting database (this value gets periodically updated as a user submits jobs over time)
- **urgency**: represents the user's urgency when submitting a job (default value is 16)

As additional factors get added (**job size**, **age**, **queue**, etc.), they will be added to this equation to generate a more representative multi-factor priority. 

#### Implementation

This PR adds a new folder to `flux-accounting/src/` called `plugins/` which stores both the priority plugin itself and the external service that pushes data to the plugin.

**bulk_update.py** is the external service that pushes user accounting data from a flux-accounting DB to the priority plugin, where it is stored in a multimap data structure to be looked up when a job priority needs to be calculated.

It is a Python script that can be run from the command line:

```console
[moussa1@docker-desktop plugins]$ flux python bulk_update.py -h
usage: bulk_update.py [-h] [-p PATH]

Description: Send a bulk update of user information from a flux-accounting
database to the multi-factor priority plugin.

optional arguments:
  -h, --help            show this help message and exit
  -p PATH, --path PATH  specify location of database file
```

It has one optional argument, `--path`, a path to a flux-accounting DB. If no `--path` is specified, it will look for a default location of the flux-accounting DB, similar to the front-end Python command line tool **flux-account.py**.

**bulk_update.py** runs one query to the flux-accounting DB to fetch every user row. Specifically, it queries the username and their respective fairshare value. It packs this data in a dictionary object and sends an RPC request to the plugin. 

Just before it begins sending user data to the plugin, it sends an empty message to the plugin, telling it to clear its multimap's contents to prepare for an update.


**mf_priority.cpp** is the priority plugin file. It contains a number of callbacks and a helper function that is responsible for calculating the priority itself.

`clear_map_cb ()` and `rec_update_cb ()` are the functions responsible for clearing and updating the multimap data structure in the plugin and interacting with the external service, **bulk_update.py**.

When a job gets submitted, `validate_cb ()` runs and performs a lookup in the multimap to look for the userid that submitted the job. If the userid cannot be found, the job is rejected with an error message explaining that the user couldn't be found in the flux-accounting DB.

If a user is indeed in the flux-accounting DB, the `userid` and `urgency` are extracted using `flux_plugin_arg_unpack ()` and sent to a helper function `priority_calculation ()`, where its priority is calculated and returned, to be packed using `flux_plugin_arg_pack ()`.

`priority_calculation ()` is responsible for calculating a job's priority using the formula described in the _Background_. It currently covers two corner cases:

1. If a user submits a job with an urgency of `0`, the priority returned is `FLUX_JOB_PRIORITY_MIN` = 0. A job will remain as pending.

2. If a job is submitted (or updated) with a max urgency of `31`, the priority returned is `FLUX_JOB_PRIORITY_MAX` = 4294967295. This is the highest priority a job can have. 

If the submitted job is neither of these, then the calculation proceeds normally with the formula stated above. I think a note should be made that the calculated priority is an absolute value. I chose this in case the priority calculated turns out to be a negative value (I think the scenario in which this happens is if a user's fairshare value is so low that, when subtracted from the urgency, it results in a small negative integer; taking an absolute value _should_ result in a very low integer priority).

#### Examples

<details>

Since this PR does not have testing yet, I thought it might be beneficial to provide at least a couple examples to show how the plugin works in its current state.

Here is how I compile the plugin, load it, and push data to it:

```console
[moussa1@docker-desktop plugins]$ g++ -fPIC -shared -o mf_priority.so mf_priority.cpp
[moussa1@docker-desktop plugins]$ flux jobtap load mf_priority.so
Loaded:
mf_priority
Previously loaded:
builtin.priority.default
[moussa1@docker-desktop plugins]$ flux python bulk_update.py -p test/accounting_db_data/small_no_tie.db
```

* submitting a job with no specification of urgency:

```console
[moussa1@docker-desktop plugins]$ flux mini submit -n 1 hostname
ƒZ8pD9HtK
[moussa1@docker-desktop plugins]$ flux job wait-event -v ƒZ8pD9HtK priority
1622047983.174637 submit userid=58985 urgency=16 flags=0
1622047983.190122 depend
1622047983.190164 priority priority=45321
```

* submitting a job with `--urgency=15`

```console
[moussa1@docker-desktop plugins]$ flux mini submit -n 1 --urgency=15 hostname
ƒZuN5hafZ
[moussa1@docker-desktop plugins]$ flux job wait-event -v ƒZuN5hafZ priority
1622048084.122513 submit userid=58985 urgency=15 flags=0
1622048084.137835 depend
1622048084.137883 priority priority=45320
```

* submitting a job with `--urgency=0` and updating it later with `--urgency=31`

```console
[moussa1@docker-desktop plugins]$ flux mini submit -n 1 --urgency=0 hostname
ƒaQNfiUvB
[moussa1@docker-desktop plugins]$ flux job wait-event -v ƒaQNfiUvB priority
1622048149.879270 submit userid=58985 urgency=0 flags=0
1622048149.894008 depend
1622048149.894067 priority priority=0
[moussa1@docker-desktop plugins]$ flux job urgency ƒaQNfiUvB 31
[moussa1@docker-desktop plugins]$ flux job wait-event -v ƒaQNfiUvB start
1622048149.879270 submit userid=58985 urgency=0 flags=0
1622048149.894008 depend
1622048149.894067 priority priority=0
1622048183.573533 urgency userid=58985 urgency=31
1622048183.573644 priority priority=4294967295
1622048183.579611 alloc annotations={"sched":{"resource_summary":"rank0/core0"}}
1622048183.588152 start
```
</details>

#### Assumptions/Gaps/Testing

<details>

I'll use this section to document the gaps for this PR I can think of so far, the assumptions made, and any other questions or feedback I think I could use.

* Testing

~~This is one of the larger gaps I can think of for this PR. I looked at the tests for the plugins in flux-core and am wondering how I should approach testing in the flux-accounting repo. Are sharness tests a good approach for my plugin?~~ 

~~Maybe I could populate a fake database with some fake users and fairshare values and have them submit similar jobs and make sure their calculated priorities are as expected, but I guess I wasn't sure how I would do this if I am not in a flux instance.~~ 

* Only Using User/Fairshare

~~One of the assumptions I've made so far in this PR is that a user only belongs to one bank; with this assumption, bank information doesn't need to be passed from the DB to the plugin. However, users can belong to multiple banks, so I need to handle this case.~~ 

* Clearing the map for an update

In its current structure, a job could be submitted at the time just after the map data structure gets cleared and just before an update is pushed to the plugin. In this case, the job would be rejected saying that the user wasn't found in the flux-accounting DB. Would the correct way to handle this would be to instead `return flux_jobtap_priority_unavail ()` so that the submitted job will remain in the `PRIORITY` state until a priority is assigned using `flux_jobtap_reprioritize_job ()`?

* Declaring the map as a global variable

The plugin declares the map as a global variable. I declared it as global because multiple functions and callbacks interact with and update this map. Is this okay behavior? Should it instead be declared and initialized in one of the callbacks and passed to other functions when it is needed?

These are the `TODOS` and clarifications I can think of so far, but maybe there are things I missed. 
</details>

#### PR Structure

I think this PR can be broken into the following parts: 

- Part One adds a `userid` field to the flux-accounting DB. 
- Parts Two and Three add the priority plugin and the external bulk update service respectively.
- Part Four adds some sharness tests for the plugin as well and adds some helper scripts required by the new sharness test (e.g. add an `rc` directory)

grondo commented:
> This is one of the larger gaps I can think of for this PR. I looked at the tests for the plugins in flux-core and am wondering how I should approach testing in the flux-accounting repo. Are sharness tests a good approach for my plugin?

> Maybe I could populate a fake database with some fake users and fairshare values and have them submit similar jobs and make sure their calculated priorities are as expected, but I guess I wasn't sure how I would do this if I am not in a flux instance.

I would definitely start with a sharness test. If you have not already, pull in sharness.sh and `sharness.d/flux-sharness.sh` from flux-core. You can then use the `test_under_flux` "function" to re-run sharness test(s) under a flux instance.

There are a few existing priority-order tests you can look to as examples in flux-core:
```
t2216-job-manager-priority-order-single.t
t2217-job-manager-priority-order-limited.t
t2218-job-manager-priority-order-unlimited.t
```

You could also easily create your fake database, and/or have bulk_update.py push contrived values into your plugin, etc. Basically anything you can script can be tested.

cmoussa1 commented:
OK, I've updated this PR with the following changes:

- added handling for users who belong to more than one bank
- added sharness tests that load the plugin and submit jobs using the custom plugin

So I'll go ahead and bring this PR out of [WIP] and post it for official review. I handled the multiple-bank case in the following way:

When a user submits a job, the validate callback will try to unpack a bank from the jobspec and store it in a `bank` variable if it finds one. If not, `bank` is left NULL. In the case where `bank` is left `NULL` but the user submitting the job belongs to more than one bank, the validate callback will reject the job, explaining that the user must specify a bank to charge since they belong to more than one bank. 

Eventually, I think the plugin should handle this case by not rejecting the job and just using a default bank (e.g setting the default to the first bank the user was added to in the DB), and I'll open an issue to track this so I can add it in the near future.

grondo commented:
> Eventually, I think the plugin should handle this case by not rejecting the job and just using a default bank (e.g setting the default to the first bank the user was added to in the DB), and I'll open an issue to track this so I can add it in the near future.

This sounds fine for now. I would think eventually administrators would want to be able to tweak the default bank for a user outside of the order in which the user is added to banks, so I would consider making the default bank a property of each user which could be separately set in the db.

> Oh, one other thing I wanted to make sure I was doing correctly was compiling the plugin into a .so using Makefile.am. I think I came up with a solution in this PR, as a .libs directory containing the .so gets created during make. Is this the correct way to compile the plugin?

You got most of the way there! You'll want to also add `-avoid-version` to avoid adding a DSO version to the plugin (e.g. it is now called `mf_priority.so.0.0.0`, and use `-export-symbols-regex '^flux_plugin_init\$\$'` so that only the `flux_plugin_init()` symbol is exported from the DSO.

You can just copy the `fluxplugin_ldflags` from the `configure.ac` in flux-core:`
```
fluxplugin_ldflags="-avoid-version -export-symbols-regex '^flux_plugin_init\$\$' --disable-static -shared -export-dynamic"
AC_SUBST(fluxplugin_ldflags)
```

And then use that on your plugin's LDFLAGS.

grondo commented on src/plugins/mf_priority.cpp:
```diff
@@ -0,0 +1,271 @@
+    // search element in multimap by key
+    it = users.find (userid);
+    if (it == users.end ())
+        return flux_jobtap_reject_job (p, args,
+                                       "user not found in flux-accounting DB");
```
`flux_jobtap_reject_job()` can only be used from `job.validate` (or `job.dependency.*`), but `priority_calculation()` looks like it can be called from `job.priority.get`. In this case, you can return `flux_jobtap_priority_unavail()` if the job is still in PRIORITY state. It looks like this can't happen though -- if the user does not exist in the db then the job would have been rejected by `validate_cb`.

cmoussa1 replied:
Ah, yes I think you're right - `validate_cb ()` would have rejected the job first if the user does not exist in the DB, so I don't think there is a need for the `flux_jobtap_reject_job ()` in `priority_calculation ()`; I can remove that.

grondo reviewed (changes requested):
@cmoussa1, I made a first pass with some comments. I mainly reviewed the multifactor plugin, since I'm not the right person to review the remainder of the code.

I don't understand enough of how the user+bank is used to get the fairshare factor to know if the use of a multimap is the right data structure here, it seems like it requires a little extra work currently. 

It may be easier if you had two maps -- one that maps banks to fairshare value, and one that maps userid to a list of banks in which the user is a member. Then you can easily test bank membership, and get the bank fairshare value in a single lookup call.

I'd also suggest caching the bank along with the job, perhaps with `flux_jobtap_job_aux_set()` so that you don't have to continually lookup the bank name for each job when you've already done it in the `job.validate` callback.

I could be misunderstanding the role of the users map though, so just correct me if I'm wrong.

cmoussa1 commented on src/plugins/mf_priority.cpp:
```diff
@@ -0,0 +1,271 @@
+}
+
+
+char v[] = "job.validate";
+char p[] = "job.state.priority";
+char g[] = "job.priority.get";
```
Thanks for pointing this out - the reason I defined them is because C++ gave me a warning about converting strings to `char *` if I just use the strings directly in `tab[]`:

```console
warning: ISO C++ forbids converting a string constant to ‘char*’
```

Is it OK to ignore this warning?

grondo replied:
Ah, this is because the topic string in `struct flux_plugin_handler` is `char *topic` instead of `const char *topic`. This is a bug in the `flux_plugin_t` interface. I'll open a bug in flux-core and fix it, and then those warnings should be resolved.

Sorry about that!

cmoussa1 replied:
No worries! Thanks for helping me walk through that. 🙂

grondo replied:
A fix is up for this at flux-framework/flux-core#3720

cmoussa1 commented on src/plugins/mf_priority.cpp:
```diff
@@ -0,0 +1,271 @@
+
+    // searching element in std::map by key
+    it = users.find (userid);
+    if (it == users.end ())
+        return flux_jobtap_reject_job (p, args,
```
Thank you for the reminder - I'll go ahead and add a test for this.

cmoussa1 commented:
Thanks for your feedback so far @grondo - maybe it makes sense to continue talking out my design decision regarding the data structure used in the plugin before I make more changes.

> It may be easier if you had two maps -- one that maps banks to fairshare value, and one that maps userid to a list of banks in which the user is a member. Then you can easily test bank membership, and get the bank fairshare value in a single lookup call.
>
> I'd also suggest caching the bank along with the job, perhaps with flux_jobtap_job_aux_set() so that you don't have to continually lookup the bank name for each job when you've already done it in the job.validate callback.
>
> I could be misunderstanding the role of the users map though, so just correct me if I'm wrong.

So the idea of using a multimap data structure was to make use of the ability to have multiple entries with the same key in the data structure, i.e there could potentially be two entries in the multimap that have a key of `userA`, reflecting the scenario where a user who belongs to two banks `bank1` and `bank2`, each with their own respective fairshare values. 

```
key      value
userA -> <'bank1', 0.12345>
userA -> <'bank2', 0.54321>
.
.
.
userB -> <'bank1', 0.98765>
```

Using a multimap is the reason I needed to implement a `clear_map_cb ()`, because if I didn't clear it, old entries wouldn't be overwritten during an update. I hope this explanation makes sense. 

Maybe it would be better to use your suggestion and use two maps, although I am trying to think of how to look up an individual user's fairshare value just using the bank they belong to.

grondo commented:
> Maybe it would be better to use your suggestion and use two maps, although I am trying to think of how to look up an individual user's fairshare value just using the bank they belong to.

Ah, does each user,bank pair have a different fairshare value? Sorry, I'm not familiar with the overall architecture here so I probably shouldn't comment on specific design decisions!

> Using a multimap is the reason I needed to implement a clear_map_cb (), because if I didn't clear it, old entries wouldn't be overwritten during an update. I hope this explanation makes sense.

Yes, however, I was mainly pointing out that having separate "clear" and "update" RPCs makes the plugin racy, since a number of things could happen in between the two RPCs (e.g. a job submission), whereas if you perform the clear in the update rpc, then you are safe from races since the job manager module/thread cannot do any asynchronous work while in your callback.

Does that make sense? I would just combine the separate clear and update service callbacks into a single clear+update callback. You could make the clear optional in the future if necessary by adding a `clear` boolean flag to your update RPC.

cmoussa1 commented:
> Ah, does each user,bank pair have a different fairshare value? Sorry, I'm not familiar with the overall architecture here so I probably shouldn't comment on specific design decisions!

No, I appreciate you asking these questions and helping me think through these things! Yes, user-bank pairs should each have their own fairshare values.

> Yes, however, I was mainly pointing out that having separate "clear" and "update" RPCs makes the plugin racy, since a number of things could happen in between the two RPCs (e.g. a job submission), whereas if you perform the clear in the update rpc, then you are safe from races since the job manager module/thread cannot do any asynchronous work while in your callback.
>
> Does that make sense? I would just combine the separate clear and update service callbacks into a single clear+update callback. You could make the clear optional in the future if necessary by adding a clear boolean flag to your update RPC.

And yeah, I think this makes sense. I should combine the two into just one clear-and-update RPC. 

I believe I'll need to think about how the `clear ()` occurs in this callback, because as of now, the external service Python script sends one RPC for every user-bank pair, so if I only put a `clear ()` at the top of the update RPC, the map will get cleared every time a payload gets sent and will only contain the most recent JSON payload.

grondo commented:
> I believe I'll need to think about how the clear () occurs in this callback, because as of now, the external service Python script sends one RPC for every user-bank pair, so if I only put a clear () at the top of the update RPC, the map will get cleared every time a payload gets sent and will only contain the most recent JSON payload.

It will be much better if all updates are sent in a single RPC if we can figure out how to do that. The mass of RPCs for a large system will be inefficient, plus the race condition you've noted with clearing the multimap first.

One thing you could do is instead of using a multimap, use a map of maps (or hash of hashs). The outer map is would be by userid, and the inner map would be by bank name. This would allow you to _update_ existing bank entries instead of clearing and re-adding them. Then the bulk-update tool could send a single RPC with a JSON object which contained all updated user/bank pairs, and the update RPC would just iterate over those and do the updates.

cmoussa1 commented:
> One thing you could do is instead of using a multimap, use a map of maps (or hash of hashes). The outer map is would be by userid, and the inner map would be by bank name. This would allow you to update existing bank entries instead of clearing and re-adding them. Then the bulk-update tool could send a single RPC with a JSON object which contained all updated user/bank pairs, and the update RPC would just iterate over those and do the updates.

Yup, absolutely. I don't think a `clear_map` callback would be required at all if we use a map of maps. And I think accessing values from this map of map feels more straightforward, so this might be the way to go.

The one drawback that I believe comes with using a map of maps arises when trying to access a "default" bank for a user. I think C++ maps are internally ordered by the key, so defaulting to the first bank found under the `userid` key could potentially change when new banks are added. 

I wonder if adding a field in the user table denoting which bank should be treated as the default bank would be a good way to handle this. I could then send that data in the external update service, and in the internal map where the key is `bank`, I could lookup this default boolean value to determine if the bank is the default one.

grondo commented:
> I wonder if adding a field in the user table denoting which bank should be treated as the default bank would be a good way to handle this. I could then send that data in the external update service, and in the internal map where the key is bank, I could lookup this default boolean value to determine if the bank is the default one.

Or, you could reserve the bank name "default" and store a copy or reference of the default bank under `users[userid]["default"]`.

cmoussa1 commented:
Thanks for the helpful review so far - I've just pushed up a round of `fixup` commits that I hope address the feedback given:

- `fluxplugin_ldflags` was copied over from flux-core for compiling the plugin
- the data structure in the plugin was changed from a `multimap` to a `map` of `map`s; the outer layer has a key of `userid`, and the inner layer has a key of `bank` and respective `fairshare` value in that bank. A `"default"` bank entry is also created for the user when they are added to the plugin's data structure. The `"default"` bank uses the first bank the user was added to in the flux-accounting database. I think a TODO for this PR would be to add support for changing `"default"` banks both in the flux-accounting database and in the plugin itself.
- Because of the change to the data structure used in the plugin ⬆️, the `clear_map` callback is no longer needed, since the map can simply be updated with new values without having to clear out the keys first.
- `flux_jobtap_reject_job ()` was removed from `job.priority.get` since the plugin already checks for the user's existence in the flux-accounting database during the `job.validate` state. 
- Two new tests were added: one to confirm that a job would be rejected in the `job.validate` state if they did not belong in the flux-accounting DB, and one that uses the default bank when submitting a job

cmoussa1 commented:
Just squashed my `fixup` commits and re-pushed.

grondo commented on t/t1001-mf-priority.t:
```diff
@@ -0,0 +1,90 @@
+test_expect_success 'submit a job with default urgency' '
+        jobid=$(flux mini submit --setattr=system.bank=account3 -n1 hostname) &&
+        flux job wait-event -f json $jobid priority | jq '.context.priority' > test_job1.txt &&
+        test_cmp test_job1.txt ${FLUX_BUILD_DIR}/t/expected/job_1_priority.txt
```
Instead of storing expected output separately from the test under version control, it is probably easier and more maintainable to create the expected output within the test itself (you'll see the pattern quite often in flux-core tests), e.g. here:
```
test_expect_success 'submit a job with default urgency' '
￼        jobid=$(flux mini submit --setattr=system.bank=account3 -n1 hostname) &&
￼        flux job wait-event -f json $jobid priority | jq '.context.priority' > test_job1.txt &&
         cat >job1.expected <<-EOF
         4096
         EOF
￼        test_cmp job1.expected test_job1.txt
```

Also note that usage of `test_cmp` is `test_cmp expected actual`.

I would use this method for almost all tests unless the expected output is very large.

cmoussa1 replied:
Thanks for pointing this out. I've just pushed a change to use the above pattern to compare the test result to the expected value. This feels much cleaner than the approach I was originally using.

grondo commented on t/t1001-mf-priority.t:
```diff
@@ -0,0 +1,90 @@
+        flux jobtap load -r .priority-default ${MULTI_FACTOR_PRIORITY}
+'
+
+test_expect_success 'list all loaded plugins' '
+        flux jobtap list -a > test.txt
+'
+
+test_expect_success 'make sure multi-factor priority plugin is loaded' '
+        test_cmp test.txt ${FLUX_BUILD_DIR}/t/expected/loaded_priority.txt
+'
```
Suggestion: do not use `test_cmp` to test output of `flux jobtap list` because you never know, some default plugin besides just your priority plugin could be loaded. Instead, combine these two tests into one and use `grep` to ensure your plugin was loaded:

```
test_expect_success 'make sure multi-factor priority plugin is loaded' '
        flux jobtap list | grep mf_priority
'
```

Remove `loaded_priority.txt` from version control

cmoussa1 replied:
Ah, good point. Thank you. I've just pushed up a change in the test to just use `grep` to make sure my plugin is loaded.

grondo reviewed (commented):
Just a quick few comments on the tests. I'd remove all "expected output" from version control and generate or check expected output within the tests (see comments inline)

cmoussa1 commented:
Just pushed to catch up after #124 and #126.

grondo commented on src/plugins/mf_priority.cpp:
```diff
@@ -0,0 +1,230 @@
+                             "userid", &uid,
+                             "bank", &bank,
+                             "fairshare", &fshare) < 0) {
+        flux_log_error (h, "failed to unpack custom_priority.trigger msg");
```
It would probably be good to add a test for this error condition. Just send an empty payload or other invalid input to the mf_priority plugin's service and ensure that the RPC fails expectedly, and that the job manager or instance doesn't crash ;-)

cmoussa1 replied:
Does sending an empty or invalid input payload causes an error to happen just before it gets to the `flux_log_error` message? For example, if an empty payload is sent:

```
Traceback (most recent call last):
  File "bad_payload.py", line 8, in <module>
    flux.Flux().rpc("job-manager.mf_priority.rec_update", {}).get()
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 78, in get
    resp_str = self.get_str()
  File "/usr/lib64/flux/python3.6/flux/util.py", line 76, in func_wrapper
    retval = func(future, *args, **kwargs)
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 72, in get_str
    self.pimpl.flux_rpc_get(payload_str)
  File "/usr/lib64/flux/python3.6/flux/util.py", line 52, in func_wrapper
    raise EnvironmentError(error.errno, errmsg.decode("utf-8")) from None
OSError: [Errno 71] Object item not found: userid
```

Or if an invalid payload gets sent:

```
Traceback (most recent call last):
  File "bad_payload.py", line 8, in <module>
    flux.Flux().rpc("job-manager.mf_priority.rec_update", data).get()
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 78, in get
    resp_str = self.get_str()
  File "/usr/lib64/flux/python3.6/flux/util.py", line 76, in func_wrapper
    retval = func(future, *args, **kwargs)
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 72, in get_str
    self.pimpl.flux_rpc_get(payload_str)
  File "/usr/lib64/flux/python3.6/flux/util.py", line 52, in func_wrapper
    raise EnvironmentError(error.errno, errmsg.decode("utf-8")) from None
OSError: [Errno 71] Expected string, got integer
```

I could very well be doing something wrong though. Is there a proper way to handle this failed unpack when a payload is empty or of invalid format?

grondo replied:
That looks like the expected result to me, since you send the result of `flux_msg_last_error()` in the error response. You may have to run `flux dmesg` to see the logged error.

cmoussa1 replied:
Ah, okay, thanks for pointing that out. I haven't used `flux dmesg` before, so apologies if I am totally using it wrong. I just pushed a `--fixup` commit (6c27d7a) that adds a test that uses `flux dmesg` to make sure that `flux_log_error` reports the right message.

grondo replied:
LGTM. I think you are using it correctly. In this case, you are getting an error back from the RPC which indicates the incorrect payload was caught by your plugin, so IMO it would be sufficient to ensure the `bad_payload.py` script fails with an expected error message.

grondo commented on src/plugins/mf_priority.cpp:
```diff
@@ -0,0 +1,230 @@
+
+    // searching element in std::map by key
+    it = users.find (userid);
+    if (it == users.end ())
```
Should `job.validate` also check that the `bank`, if provided, is valid for the given user?

cmoussa1 replied:
Yes, you're right. I just pushed up a `fixup` commit (e92974b) that adds another check in `validate_cb ()` where the passed-in bank is checked under the user, and if it is not found, the job will be rejected with an error message saying that the user does not belong to the passed-in bank. I also added a test (4cc116c) that should trigger this error message.

grondo reviewed (changes requested):
This is looking good! Just a few pretty minor comments in line. This time I focused on the job manager plugin specifically.

I think this is very close to ready for a merge :+1:

cmoussa1 commented on src/plugins/mf_priority.cpp:
```diff
@@ -0,0 +1,230 @@
+{
+    flux_plugin_t *p;
+    assert (sizeof arg == sizeof p);
+    memcpy (&p, &arg, sizeof arg);
```
Thanks for letting me know about this. Just using:

```
flux_plugin_t *p = arg;
```

was giving me an error; I _think_ it was because the C++ compiler complains about an implicit conversion type to `void *`. I referenced [this](https://stackoverflow.com/questions/30088025/malloc-invalid-conversion-from-void-to-double) SO post. Let me know if I am way off base here. 

Regardless, I dropped those lines because, like you mentioned, they are not used anywhere in the function.

grondo replied:
Ah, sorry, C++ must be less forgiving of type conversion. You would want to do something like this instead:
```C
    flux_plugin_t *p = (flux_plugin_t *) arg;
```

Though I didn't test that with a c++ compiler so I could be wrong again 😉

cmoussa1 commented on src/plugins/mf_priority.cpp:
```diff
@@ -0,0 +1,230 @@
+
+    char *uid, *fshare, *bank;
+
+    if (flux_request_unpack (msg, NULL, "{s:s, s:s, s:s}",
```
Ah, I don't have an issue open on this, but I should open one so I can track it. Maybe I can work on this along with the multi-user testing for the planned July release.

cmoussa1 commented on src/plugins/mf_priority.cpp:
```diff
@@ -0,0 +1,230 @@
+                                "userid", &userid,
+                                "jobspec", "attributes", "system",
+                                "bank", &bank) < 0) {
+        flux_log (h,
```
This sounds like a good idea. I pushed up a `fixup` commit (0b80bd2) that just rejects the job if unpacking the bank fails instead of logging an error. I also added a test (c9d2587) that should trigger this rejection message when an invalid format is passed in, like your example above where `setattr=system.bank=1`.

grondo commented:
Looks good! Feel free to squash all the fixup commits and I can give the PR one more pass.

cmoussa1 commented:
Thanks! Just squashed my `fixup` commits and re-pushed.

grondo commented on src/plugins/bulk_update.py:
```diff
@@ -0,0 +1,65 @@
+import flux
```
Should this file have a copyright header? I'm actually unsure, but might as well throw it in here just in case.

cmoussa1 replied:
Oh yeah, thank you for pointing this out. I just pushed up a fix that adds a copyright header. Sorry for forgetting to put one. 🤦‍♂️

grondo commented on src/test/docker/bionic/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM fluxrm/flux-core:bionic
+FROM fluxrm/flux-core:bionic-v0.27.0
```
Update to v0.28.0 tag?

cmoussa1 replied:
Yup, right on time with flux-core! Just pushed up a change that updates this to `v0.28.0`.

grondo commented on t/t1001-mf-priority.t:
```diff
@@ -0,0 +1,127 @@
+        flux jobtap load -r .priority-default ${MULTI_FACTOR_PRIORITY}
+'
+
+test_expect_success 'list all loaded plugins' '
```
```suggestion
test_expect_success 'check that mf_priority plugin is loaded' '
```

grondo commented on t/t1001-mf-priority.t:
```diff
@@ -0,0 +1,127 @@
+#create a JSON payload
+flux.Flux().rpc("job-manager.mf_priority.rec_update", {}).get()
+EOF
+flux python bad_payload.py
```
Check that `bad_payload.py` fails with `test_must_fail`.

Also, missing `&&` between these two commands. (You might try running tests with `--chain-lint` to see if there are any other missing `&&`-chains.)


```suggestion
test_must_fail flux python bad_payload.py &&
```

grondo commented on t/t1001-mf-priority.t:
```diff
@@ -0,0 +1,127 @@
+test_expect_success 'submit a job with default urgency' '
+        jobid=$(flux mini submit --setattr=system.bank=account3 -n1 hostname) &&
+        flux job wait-event -f json $jobid priority | jq '.context.priority' > job1.test &&
+        cat <<-EOF >job1.expected
```
missing `&&`
```suggestion
        cat <<-EOF >job1.expected &&
```

grondo commented on t/t1001-mf-priority.t:
```diff
@@ -0,0 +1,127 @@
+test_expect_success 'submit a job with custom urgency' '
+        jobid=$(flux mini submit --setattr=system.bank=account3 --urgency=15 -n1 hostname) &&
+        flux job wait-event -f json $jobid priority | jq '.context.priority' > job2.test &&
+        cat <<-EOF >job2.expected
```
```suggestion
        cat <<-EOF >job2.expected &&
```

cmoussa1 replied:
It looks like `--chain-lint` fails unexpectedly when appending `&&` after these lines (`cat <<-EOF > job.expected`), but doesn't fail when I leave them out. The most recently pushed of `t1001-mf-priority.t` passes with `--chain-lint` though:

```
[moussa1@docker-desktop t]$ ./t1001-mf-priority.t --chain-lint
ok 1 - load multi-factor priority plugin
ok 2 - check that mf_priority plugin is loaded
ok 3 - try to submit a job when user does not exist in DB
ok 4 - send an empty payload to make sure unpack fails
ok 5 - create fake_payload.py
ok 6 - update plugin with sample test data
ok 7 - submit a job with default urgency
ok 8 - submit a job with custom urgency
ok 9 - submit a job with urgency of 0
ok 10 - submit a job with urgency of 31
ok 11 - submit a job with other bank
ok 12 - submit a job using default bank
ok 13 - submit a job using a bank the user does not belong to
ok 14 - reject job when invalid bank format is passed in
# passed all 14 test(s)
1..14
```

grondo replied:
Hm, sorry about the apparent misinformation, but let me check into this.

cmoussa1 replied:
No worries - it could definitely be something I'm doing wrong. I was pretty confused that `--chain-lint` was complaining about the following line:

```
cat <<-EOF >bad_payload.py &&
```

which, if no `&&` is written, dumps out this error:

```
error: bug in the test script: broken &&-chain: 
cat <<-EOF >bad_payload.py
import flux
```

but not 

```
cat <<-EOF >job1.expected
```

which, if `&&` is written, dumps out this:

```
FATAL: Unexpected exit with code 2
flux-start: 0 (pid 16667) exited with rc=1
```

grondo replied:
Ok, the problem at least in some of these cases is that you are using 8 spaces instead of tabs.
Specifically, `<<-EOF` tells the shell to ignore TABS up to the next `<tab>EOF`. If you convert to tabs (which I think we've settled on for sharness tests) then I think you'll be able to add the `&&` to the `cat` lines.

The problem with leaving this `&&` off is that the commands before the `cat <<-EOF` could fail, but the test itself could succeed.

cmoussa1 replied:
Ahhh, okay. Yeah, I must have been converting tabs to spaces when I was saving the `.t` in my IDE. I'm so sorry for the confusion! The most recent push should have the `&&`'s added to the sharness tests.

cmoussa1 replied:
That would also probably explain why it would complain about an unexpected indent when creating the Python scripts inside the sharness tests. I should push up and squash up a fix that properly indents those sharness tests.

grondo replied:
yeah, the indents between the `cat <<-EOF` and the `EOF` have to be tabs. I think without that the `cat` will read all the way to the closing quote of the `test_expect_success()` -- so you aren't actually running the test! :-)

cmoussa1 replied:
Oh my goodness 😳, well, thank you for catching that and informing me why that happens. This latest push should have replaced the spaces with tabs to prevent the tests from not running.

grondo reviewed (commented):
Some more very minor comments this round.

You should ensure that `./t1001-mf-priority --chain-lint` runs correctly. I think there are some missing `&&` after uses of `cat` in most of the tests (an easy fix) I just called out a couple cases.

grondo commented on t/t1001-mf-priority.t:
```diff
@@ -0,0 +1,127 @@
+flux setattr log-stderr-level 1
+
+test_expect_success 'load multi-factor priority plugin' '
+		flux jobtap load -r .priority-default ${MULTI_FACTOR_PRIORITY}
```
Hm, now all the indents seem to be 2 tabs? 
```suggestion
	flux jobtap load -r .priority-default ${MULTI_FACTOR_PRIORITY}
```

cmoussa1 replied:
Oops, I had always had the indents at 2 tabs. These tests should now all be indented at one tab (4 spaces).

grondo replied:
Yeah, it is confusing. I have my editor set to 8 space tabs (at least for the sharness tests), so for me the lines were indented 16 spaces! A single tab was probably plenty. :slightly_smiling_face:

grondo reviewed (approved):
This LGTM! Feel free to set MWP when ready as far as I'm concerned. Thanks for all the work!

cmoussa1 commented:
Thanks @grondo for all of your feedback and help!

cmoussa1 commented:
Oh, I think this might have to be manually merged since I made a change to `.github/workflows/main.yml`.

grondo commented:
Ah yes, sorry I missed that. I'll go ahead and press the merge button.

cmoussa1 commented:
Thank you!
